### PR TITLE
EN6-78 Content Types

### PR DESCRIPTION
**Pull request by jeffgo10**
~~#10 is required to be approved and merged before going into this PR.~~

This is the entire content type CRUD for CMS. Mostly these are derived from Data Types section in App Builder. Although there are some areas that are modified, including:
- global `history` object are not used in thunks but rather being used in container components.
- `formattedText` calls have been dropped - replaced with `react-intl` formatMessage calls.
- unit tests increased to reach at least 90% unit test coverage

### CODE REVIEW

File: src/state/pagination/types.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// eslint-disable-next-line
```
Comment (ichalagashvili) [inline src/state/pagination/types.js]: Specify what rule you are disabling or adjust the code to export default const

File: src/ui/common/attributes/AttributeRole.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import { FormattedMessage } from 'react-intl';
+import { Row, Col, FormGroup } from 'patternfly-react';
+import { FieldArray } from 'redux-form';
+import RoleSelectRenderer from 'ui/common/form/RoleSelectRenderer';
+
+class AttributeRole extends Component {
+  componentWillMount() {
+    const { onWillMount, ...allprops } = this.props;
+    onWillMount(allprops);
+  }
```
Comment (ichalagashvili) [inline src/ui/common/attributes/AttributeRole.js]: Maybe change to componentDidMount() ?

https://daveceddia.com/where-fetch-data-componentwillmount-vs-componentdidmount/

File: src/ui/common/form/AttributeForm.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  TYPE_LIST,
+  TYPE_NUMBER,
+  TYPE_THREESTATE,
+  TYPE_TIMESTAMP,
+} from 'state/content-type/const';
+
+export class AttributeFormBody extends Component {
+  componentDidMount() {
+    const { onWillMount } = this.props;
+    onWillMount();
```
Comment (ichalagashvili) [inline src/ui/common/form/AttributeForm.js]: It is just naming but seems illogical though, as onWillMount is called in componentDidMount

File: src/ui/common/form/AttributeForm.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        <Row>
+          <Col xs={12}>
+            <Button
+              className="pull-right AttributeForm__continue--btn"
+              type="submit"
+              bsStyle="primary"
+              disabled={invalid || submitting}
+            >
+              {
+                isEditComposite || isAddAttributeComposite ? <FormattedMessage id="cms.label.save" /> : <FormattedMessage id="cms.label.continue" />
```
Comment (ichalagashvili) [inline src/ui/common/form/AttributeForm.js]: No ternary operators in the render method's return block. Nic told me in my PR

File: src/ui/common/form/EditAttributeForm.js [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+  TYPE_LIST,
+  TYPE_NUMBER,
+  TYPE_THREESTATE,
+  TYPE_TIMESTAMP,
+} from 'state/content-type/const';
+
+export class EditAttributeFormBody extends Component {
+  componentDidMount() {
+    const { onWillMount } = this.props;
+    onWillMount(this.props);
```
Comment (ichalagashvili) [inline src/ui/common/form/EditAttributeForm.js]: naming same here

File: src/ui/common/form/EditAttributeForm.js [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        <Row>
+          <Col xs={12}>
+            <Button
+              className="pull-right EditAttributeForm__continue--btn"
+              type="submit"
+              bsStyle="primary"
+              disabled={invalid || submitting}
+            >
+              {
+              !isComposite ? <FormattedMessage id="cms.label.continue" /> : <FormattedMessage id="cms.label.save" />
```
Comment (ichalagashvili) [inline src/ui/common/form/EditAttributeForm.js]: no ternary in return of render. Declare and assign the value above return

File: src/ui/common/form/MonolistAttributeForm.js [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+const NO_INFO_ATTRIBUTE = [
+  TYPE_BOOLEAN,
+  TYPE_CHECKBOX,
+  TYPE_THREESTATE,
+];
+
+export class MonolistAttributeFormBody extends Component {
+  componentDidMount() {
+    const { onWillMount, ...allprops } = this.props;
+    onWillMount(allprops);
```
Comment (ichalagashvili) [inline src/ui/common/form/MonolistAttributeForm.js]: same naming

File: src/ui/common/form/RenderDatePickerInput.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+class RenderDatePickerInput extends Component {
+  constructor(props) {
+    super(props);
+    this.handleChange = this.handleChange.bind(this);
+  }
+
+  componentDidMount() {
+    const { onWillMount, ...otherProps } = this.props;
+    onWillMount(otherProps);
```
Comment (ichalagashvili) [inline src/ui/common/form/RenderDatePickerInput.js]: same naming

File: src/ui/common/form/RenderDatePickerInput.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            selected={input.value ? moment(input.value, dateFormat) : null}
+            onChange={this.handleChange}
+            disabledKeyboardNavigation
+            locale={locale}
+            dateFormat={dateFormat}
+            isClearable
+            calendarClassName="RenderDatePickerInput__calendar"
+          />
+          <div className="help-block help-block-error">
+            {touched ? error : ''}
```
Comment (ichalagashvili) [inline src/ui/common/form/RenderDatePickerInput.js]: no ternary in return of render

File: src/ui/common/form/RoleSelectRenderer.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+            bsStyle="danger"
+            className="btn btn-danger RoleSelectRenderer__remove--btn"
+            onClick={() => fields.remove(i)}
+          >
+            <FormattedMessage id="cms.label.delete" />
+          </Button>
+        </Col>
+      </div>
+    ));
+  }
```
Comment (ichalagashvili) [inline src/ui/common/form/RoleSelectRenderer.js]: I had almost same like this and Nic told me that "no rendering methods should exist in the class other than default render method". He also advised me to either move this code into render and assign it to a variable or create a separate mini component and import it into render

File: src/ui/content-type/AddContentTypeForm.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+const maxLength3 = maxLength(3);
+
+const maxLength50 = maxLength(50);
+
+
+export class AddContentTypeFormBody extends Component {
+  componentDidMount() {
+    const { onWillMount } = this.props;
+    onWillMount(this.props);
```
Comment (ichalagashvili) [inline src/ui/content-type/AddContentTypeForm.js]: naming same

File: src/ui/content-type/attributes/monolist/MonolistPage.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+  ROUTE_CMS_CONTENTTYPE_LIST,
+  ROUTE_CMS_CONTENT_TYPE_ATTRIBUTE_EDIT,
+} from 'app-init/routes';
+
+import { TYPE_COMPOSITE, TYPE_MONOLIST } from 'state/content-type/const';
+
+class MonolistPage extends Component {
+  componentDidMount() {
+    const { onWillMount, ...otherProps } = this.props;
+    onWillMount(otherProps);
```
Comment (ichalagashvili) [inline src/ui/content-type/attributes/monolist/MonolistPage.js]: same naming

File: src/ui/content-type/attributes/monolist/MonolistPage.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+                </BreadcrumbItem>
+                <BreadcrumbItem to={ROUTE_CMS_CONTENTTYPE_LIST}>
+                  <FormattedMessage id="cms.contenttype.title" />
+                </BreadcrumbItem>
+                <BreadcrumbItem>
+                  <FormattedMessage id="cms.label.edit" />:
+                  &nbsp;{contentTypeCode}
+                </BreadcrumbItem>
+                {
+                  type === TYPE_COMPOSITE ? (
```
Comment (ichalagashvili) [inline src/ui/content-type/attributes/monolist/MonolistPage.js]: no ternary

**Comment by jeffgo10** [comment]
hi @ichalagashvili . Thank you for the review. Please check my changes when you can. Thanks!

**Pull request merged**